### PR TITLE
Fix temp file storage/deletion + a few new features

### PR DESCRIPTION
**Pull request by Mbucari**
- .tmp and .json files are again properly stored in DownloadsInProgressDirectory, and the are again used if resuming a download.
- Yellow stoplight works again.
- All audio files stored in DecryptInProgressDirectory are deleted if the conversion fails #144
- Added option to retain aax file #187
- Add option to remove "(Unabridged)" from metadata tags #223
- Added options for mp3 encoding
- Add option to create cue sheet
- Prevent same book being decrypted more than once at a time
- Made a new settings tab for all audio file editing/fixing options.

image

### CODE REVIEW

File: LibationWinForms/grid/GridEntry.cs [hunk 1]
```diff
@@ -77,6 +79,16 @@ public string DisplayTags
 
 		public GridEntry(LibraryBook libraryBook) => setLibraryBook(libraryBook);
 
+		public async Task DownloadBook()
+		{
+			if (!DownloadInProgress)
+			{
+				DownloadInProgress = true;
```
Comment (rmcrackan) [inline LibationWinForms/grid/GridEntry.cs:86]: I recommend wrapping this like, or similar:

    if (!DownloadInProgress)
    {
        try
        {
            DownloadInProgress = true;
            await BookLiberation.ProcessorAutomationController.BackupSingleBookAsync(LibraryBook);
        }
        finally
        {
            DownloadInProgress = false;
        }
    }

**Comment by rmcrackan** [comment]
[Dude](https://i.kym-cdn.com/photos/images/original/000/971/686/891.jpg). This is seriously impressive for 1 day's work. The hell do you need me for? I had 1 very minor comment and other than that it's ready to rock.

**Comment by Mbucari** [comment]
Ha! Thank you! When I get on a roll it's hard to stop!

I made the change you recommended, and I also added code to save the decryption key to a file if the user is saving the aax file. Cheers!

**Comment by Mbucari** [comment]
I think you're referring to the DownloadLicense constructor.  In that case, DownloadLicense was used for podcast-type audio files that were delivered as unencrypted mp3s, so null keys didn't mattwer.  Here, we only want to write the aaxc file and key file if AudiobookDownloadBase is actually downloading an aaxc file.

**Comment by rmcrackan** [comment]
> save the decryption key

Good call. I had another comment but it's just curiosity. I doubt it's blocking so I'll put this through.

I'm on the cusp of enabling library auto-update and I've got another request that should be low hanging fruit too. Once work slacks off just a tiny bit I expect to release both of those. The auto-update turned out to be surprisingly hard due to choices I made years ago about how the grid is updated -- oops. I refactored that recently, as well as significantly speeding up database import. Between your work and mine, it's going to be a *very* fine month for features.

**Comment by Mbucari** [comment]
I noticed the speed increase! It's VERY welcome. thanks!

**Comment by rmcrackan** [comment]
tl;dr: skip this entirely if you don't care about implementation details about the speed increase

This is why you benchmark. The speed increase came from the most unlikely place imaginable. In the import, I pull all relevant items into an in-memory cache that Entity Framework has then I use that. WAY faster than hitting the database: in memory (rather than hitting the db each time) and O(sqrt(n)) -- stupid fast. Except it wasn't. That step was *destroying* my benchmarks -- wtf! Turns out, those benefits only applied if you queried by a certain field which I wasn't. I already indexed by my needed fields, so hitting the db itself was twice as fast as that stupid in-mem table. Then I added a switched to using my own in-memory hashtable to mirror the database's index tables. Long story short: what was taking 20,000-40,000 ms -- after my own hashtables -- was down to ONE ms

**Comment by Mbucari** [comment]
That's crazy.  And +1 to benchmarking. TVS's built-in profiler has helped me track down several nasty performance bottlenecks!
